// @microsoft/teamsfx-core/build/common/deps-checker/constant/helpLink.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.v3NodeNotLtsHelpLink = exports.v3NodeNotSupportedHelpLink = exports.v3NodeNotFoundHelpLink = exports.vxTestAppInstallHelpLink = exports.ngrokInstallHelpLink = exports.dotnetNotSupportTargetVersionHelpLink = exports.dotnetFailToInstallHelpLink = exports.dotnetExplanationHelpLink = exports.dotnetDefaultHelpLink = exports.nodeInstallationLink = exports.nodeNotSupportedForSPFxHelpLink = exports.nodeNotSupportedForAzureHelpLink = exports.nodeNotFoundHelpLink = exports.functionDepsVersionsLink = exports.validationSettingsHelpLink = exports.v3DefaultHelpLink = exports.defaultHelpLink = void 0;
// TODO: remove this link after clean the useless code.
exports.defaultHelpLink = "https://aka.ms/teamsfx-envchecker-help";
exports.v3DefaultHelpLink = "https://aka.ms/teamsfx-actions/devtool-install";
// TODO: remove this link after clean the useless code.
exports.validationSettingsHelpLink = `${exports.defaultHelpLink}#teams-toolkit-validation-settings`;
exports.functionDepsVersionsLink = "https://aka.ms/functions-node-versions";
// TODO: remove this link after clean the useless code.
exports.nodeNotFoundHelpLink = `https://aka.ms/teamsfx-node`;
// TODO: remove this link after clean the useless code.
exports.nodeNotSupportedForAzureHelpLink = `https://aka.ms/teamsfx-node`;
// TODO: remove this link after clean the useless code.
exports.nodeNotSupportedForSPFxHelpLink = `${exports.defaultHelpLink}#nodenotsupported-spfx-hosting`;
exports.nodeInstallationLink = "https://nodejs.org/about/releases/";
exports.dotnetDefaultHelpLink = "https://aka.ms/teamsfx-actions/devtool-install";
exports.dotnetExplanationHelpLink = exports.dotnetDefaultHelpLink;
exports.dotnetFailToInstallHelpLink = exports.dotnetDefaultHelpLink;
// TODO: remove this link after clean the useless code.
exports.dotnetNotSupportTargetVersionHelpLink = `${exports.defaultHelpLink}#dotnetnotsupporttargetversion`;
// TODO: remove this link after clean the useless code.
exports.ngrokInstallHelpLink = `${exports.defaultHelpLink}#failtoinstallngrok`;
exports.vxTestAppInstallHelpLink = `${exports.defaultHelpLink}#failtoinstallteamsvideoextensibilitytestapp`;
exports.v3NodeNotFoundHelpLink = "https://aka.ms/teamsfx-node";
exports.v3NodeNotSupportedHelpLink = "https://aka.ms/teamsfx-node";
exports.v3NodeNotLtsHelpLink = "https://aka.ms/teamsfx-node";
