// @microsoft/teamsfx-core/build/error/deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureStorageSetContainerPropertiesError = exports.AzureStorageGetContainerPropertiesError = exports.AzureStorageGetContainerError = exports.AzureStorageUploadFilesError = exports.AzureStorageClearBlobsError = exports.CheckDeploymentStatusError = exports.DeployZipPackageError = exports.GetPublishingCredentialsError = exports.CheckDeploymentStatusTimeoutError = exports.DeployEmptyFolderError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../common/localizeUtils");
class DeployEmptyFolderError extends teamsfx_api_1.UserError {
    constructor(folderPath) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.DeployEmptyFolderError", folderPath),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.DeployEmptyFolderError", folderPath),
        });
    }
}
exports.DeployEmptyFolderError = DeployEmptyFolderError;
class CheckDeploymentStatusTimeoutError extends teamsfx_api_1.UserError {
    constructor(helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.CheckDeploymentStatusTimeoutError"),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.CheckDeploymentStatusTimeoutError"),
            helpLink: helpLink,
        });
    }
}
exports.CheckDeploymentStatusTimeoutError = CheckDeploymentStatusTimeoutError;
class GetPublishingCredentialsError extends teamsfx_api_1.UserError {
    constructor(appName, resourceGroup, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.GetPublishingCredentialsError", appName, resourceGroup, JSON.stringify(error) || "", "https://learn.microsoft.com/en-us/rest/api/appservice/web-apps/list-publishing-credentials#code-try-0"),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.GetPublishingCredentialsError.Notification", appName, resourceGroup),
            helpLink: helpLink,
        });
    }
}
exports.GetPublishingCredentialsError = GetPublishingCredentialsError;
class DeployZipPackageError extends teamsfx_api_1.UserError {
    constructor(endpoint, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.DeployZipPackageError", endpoint, JSON.stringify(error) || "", "https://learn.microsoft.com/azure/app-service/deploy-zip?tabs=cli"),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.DeployZipPackageError.Notification", endpoint),
            helpLink: helpLink,
        });
    }
}
exports.DeployZipPackageError = DeployZipPackageError;
class CheckDeploymentStatusError extends teamsfx_api_1.UserError {
    constructor(location, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.CheckDeploymentStatusError", location, JSON.stringify(error) || ""),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.CheckDeploymentStatusError", location, error.message || ""),
            helpLink: helpLink,
        });
    }
}
exports.CheckDeploymentStatusError = CheckDeploymentStatusError;
class AzureStorageClearBlobsError extends teamsfx_api_1.UserError {
    constructor(storageName, errorResponse, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.AzureStorageClearBlobsError", storageName, JSON.stringify(errorResponse, undefined, 4)),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.AzureStorageClearBlobsError.Notification", storageName),
            helpLink: helpLink,
        });
    }
}
exports.AzureStorageClearBlobsError = AzureStorageClearBlobsError;
class AzureStorageUploadFilesError extends teamsfx_api_1.UserError {
    constructor(storageName, localFolder, errorResponse, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.AzureStorageUploadFilesError", localFolder, storageName, JSON.stringify(errorResponse, undefined, 4)),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.AzureStorageUploadFilesError.Notification", localFolder, storageName),
            helpLink: helpLink,
        });
    }
}
exports.AzureStorageUploadFilesError = AzureStorageUploadFilesError;
class AzureStorageGetContainerError extends teamsfx_api_1.UserError {
    constructor(storageName, containerName, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.AzureStorageGetContainerError", containerName, storageName, JSON.stringify(error)),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.AzureStorageGetContainerError.Notification", containerName, storageName, error.message || ""),
            helpLink: helpLink,
        });
    }
}
exports.AzureStorageGetContainerError = AzureStorageGetContainerError;
class AzureStorageGetContainerPropertiesError extends teamsfx_api_1.UserError {
    constructor(storageName, containerName, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.AzureStorageGetContainerPropertiesError", containerName, storageName, JSON.stringify(error)),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.AzureStorageGetContainerPropertiesError.Notification", containerName, storageName, error.message || ""),
            helpLink: helpLink,
        });
    }
}
exports.AzureStorageGetContainerPropertiesError = AzureStorageGetContainerPropertiesError;
class AzureStorageSetContainerPropertiesError extends teamsfx_api_1.UserError {
    constructor(storageName, containerName, error, helpLink) {
        super({
            source: "azureDeploy",
            message: localizeUtils_1.getDefaultString("error.deploy.AzureStorageSetContainerPropertiesError", containerName, storageName, JSON.stringify(error)),
            displayMessage: localizeUtils_1.getLocalizedString("error.deploy.AzureStorageSetContainerPropertiesError.Notification", containerName, storageName, error.message || ""),
            helpLink: helpLink,
        });
    }
}
exports.AzureStorageSetContainerPropertiesError = AzureStorageSetContainerPropertiesError;
