// @microsoft/teamsfx-core/build/error/script.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScriptExecutionError = exports.ScriptTimeoutError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../common/localizeUtils");
/**
 * Script execution timeout
 */
class ScriptTimeoutError extends teamsfx_api_1.UserError {
    constructor(cmd) {
        const key = "error.script.ScriptTimeoutError";
        const errorOptions = {
            source: "script",
            name: "ScriptTimeoutError",
            message: localizeUtils_1.getDefaultString(key, cmd),
            displayMessage: localizeUtils_1.getLocalizedString(key, cmd),
        };
        super(errorOptions);
    }
}
exports.ScriptTimeoutError = ScriptTimeoutError;
/**
 * Script execution error
 */
class ScriptExecutionError extends teamsfx_api_1.UserError {
    constructor(script, message) {
        const key = "error.script.ScriptExecutionError";
        const errorOptions = {
            source: "script",
            name: "ScriptExecutionError",
            message: localizeUtils_1.getDefaultString(key, script, message),
            displayMessage: localizeUtils_1.getLocalizedString(key, script, message),
        };
        super(errorOptions);
    }
}
exports.ScriptExecutionError = ScriptExecutionError;
