// @microsoft/teamsfx-core/build/core/middleware/questionModel.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.addOfficeAddinQuestions = exports.getQuestionsForCreateProjectV2 = exports.traverseToCollectPasswordNodes = exports.QuestionModelMW = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const featureFlags_1 = require("../../common/featureFlags");
const tools_1 = require("../../common/tools");
const utils_1 = require("../../common/utils");
const constants_1 = require("../../component/constants");
const developerPortalScaffoldUtils_1 = require("../../component/developerPortalScaffoldUtils");
const spfx_1 = require("../../component/feature/spfx");
const question_1 = require("../../component/generator/officeAddin/question");
const question_2 = require("../../component/question");
const utils_2 = require("../../component/resource/appManifest/utils/utils");
const collaborator_1 = require("../collaborator");
const globalVars_1 = require("../globalVars");
const question_3 = require("../question");
/**
 * This middleware will help to collect input from question flow
 */
const QuestionModelMW = async (ctx, next) => {
    const inputs = ctx.arguments[ctx.arguments.length - 1];
    const method = ctx.method;
    let getQuestionRes = teamsfx_api_1.ok(undefined);
    if (method === "grantPermission") {
        getQuestionRes = await collaborator_1.getQuestionsForGrantPermission(inputs);
    }
    else if (tools_1.isV3Enabled() && (method === "listCollaborator" || method == "checkPermission")) {
        getQuestionRes = await collaborator_1.getQuestionsForListCollaborator(inputs);
    }
    else if (tools_1.isV3Enabled() && method === "deployAadManifest") {
        getQuestionRes = await question_3.getQuestionForDeployAadManifest(inputs);
    }
    if (getQuestionRes.isErr()) {
        globalVars_1.TOOLS === null || globalVars_1.TOOLS === void 0 ? void 0 : globalVars_1.TOOLS.logProvider.error(`[core] failed to get questions for ${method}: ${getQuestionRes.error.message}`);
        ctx.result = teamsfx_api_1.err(getQuestionRes.error);
        return;
    }
    const node = getQuestionRes.value;
    if (node) {
        const res = await teamsfx_api_1.traverse(node, inputs, globalVars_1.TOOLS.ui, globalVars_1.TOOLS.telemetryReporter);
        if (res.isErr()) {
            ctx.result = teamsfx_api_1.err(res.error);
            return;
        }
    }
    await next();
};
exports.QuestionModelMW = QuestionModelMW;
// export function desensitize(node: QTreeNode, input: Inputs): Inputs {
//   const copy = deepCopy(input);
//   const names = new Set<string>();
//   traverseToCollectPasswordNodes(node, names);
//   for (const name of names) {
//     copy[name] = "******";
//   }
//   return copy;
// }
function traverseToCollectPasswordNodes(node, names) {
    if (node.data.type === "text" && node.data.password === true) {
        names.add(node.data.name);
    }
    for (const child of node.children || []) {
        traverseToCollectPasswordNodes(child, names);
    }
}
exports.traverseToCollectPasswordNodes = traverseToCollectPasswordNodes;
async function getQuestionsForCreateProjectWithoutDotNet(inputs) {
    var _a, _b, _c, _d;
    if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
        // If toolkit is activated by a request from Developer Portal, we will always create a project from scratch.
        inputs[question_3.CoreQuestionNames.CreateFromScratch] = question_3.ScratchOptionYesVSC().id;
        inputs[question_3.CoreQuestionNames.Capabilities] =
            (_a = inputs[question_3.CoreQuestionNames.Capabilities]) !== null && _a !== void 0 ? _a : (_b = developerPortalScaffoldUtils_1.getTemplateId(inputs.teamsAppFromTdp)) === null || _b === void 0 ? void 0 : _b.templateId;
    }
    const node = new teamsfx_api_1.QTreeNode(question_3.getCreateNewOrFromSampleQuestion(inputs.platform));
    // create new
    const createNew = new teamsfx_api_1.QTreeNode({ type: "group" });
    node.addChild(createNew);
    createNew.condition = { equals: question_3.ScratchOptionYes().id };
    // capabilities
    let capNode;
    if (featureFlags_1.isPreviewFeaturesEnabled()) {
        const capQuestion = question_3.createCapabilityQuestionPreview(inputs);
        capNode = new teamsfx_api_1.QTreeNode(capQuestion);
    }
    else {
        const capQuestion = question_3.createCapabilityQuestion();
        capNode = new teamsfx_api_1.QTreeNode(capQuestion);
    }
    createNew.addChild(capNode);
    const triggerNodeRes = await question_2.getNotificationTriggerQuestionNode(inputs);
    if (triggerNodeRes.isErr())
        return teamsfx_api_1.err(triggerNodeRes.error);
    if (triggerNodeRes.value) {
        capNode.addChild(triggerNodeRes.value);
    }
    const spfxNode = await spfx_1.getSPFxScaffoldQuestion(inputs.platform);
    if (spfxNode) {
        spfxNode.condition = { equals: constants_1.TabSPFxItem().id };
        capNode.addChild(spfxNode);
    }
    // Language
    const programmingLanguage = new teamsfx_api_1.QTreeNode(question_3.ProgrammingLanguageQuestion);
    if (featureFlags_1.isPreviewFeaturesEnabled()) {
        programmingLanguage.condition = {
            notEquals: constants_1.ExistingTabOptionItem().id,
        };
    }
    else {
        programmingLanguage.condition = {
            minItems: 1,
            excludes: constants_1.ExistingTabOptionItem().id,
        };
    }
    capNode.addChild(programmingLanguage);
    // existing tab endpoint
    if (tools_1.isExistingTabAppEnabled()) {
        const existingTabEndpoint = new teamsfx_api_1.QTreeNode(question_3.ExistingTabEndpointQuestion());
        existingTabEndpoint.condition = {
            equals: constants_1.ExistingTabOptionItem().id,
        };
        capNode.addChild(existingTabEndpoint);
    }
    createNew.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
    const defaultName = !((_c = inputs.teamsAppFromTdp) === null || _c === void 0 ? void 0 : _c.appName)
        ? undefined
        : utils_1.convertToAlphanumericOnly((_d = inputs.teamsAppFromTdp) === null || _d === void 0 ? void 0 : _d.appName);
    createNew.addChild(new teamsfx_api_1.QTreeNode(question_3.createAppNameQuestion(defaultName)));
    if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
        const updateTabUrls = await getQuestionsForUpdateStaticTabUrls(inputs.teamsAppFromTdp);
        if (updateTabUrls) {
            createNew.addChild(updateTabUrls);
        }
        const updateBotIds = await getQuestionsForUpdateBotIds(inputs.teamsAppFromTdp);
        if (updateBotIds) {
            createNew.addChild(updateBotIds);
        }
    }
    // create from sample
    const sampleNode = new teamsfx_api_1.QTreeNode(question_3.SampleSelect());
    node.addChild(sampleNode);
    sampleNode.condition = { equals: question_3.ScratchOptionNo().id };
    sampleNode.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
    if (featureFlags_1.isOfficeAddinEnabled()) {
        addOfficeAddinQuestions(node);
    }
    return teamsfx_api_1.ok(node.trim());
}
async function getQuestionsForCreateProjectWithDotNet(inputs) {
    const runtimeNode = new teamsfx_api_1.QTreeNode(question_3.getRuntimeQuestion());
    const maybeNode = await getQuestionsForCreateProjectWithoutDotNet(inputs);
    if (maybeNode.isErr()) {
        return teamsfx_api_1.err(maybeNode.error);
    }
    const node = maybeNode.value;
    if (node) {
        node.condition = {
            equals: question_3.RuntimeOptionNodeJs().id,
        };
        runtimeNode.addChild(node);
    }
    const dotnetNode = new teamsfx_api_1.QTreeNode({ type: "group" });
    dotnetNode.condition = {
        equals: question_3.RuntimeOptionDotNet().id,
    };
    runtimeNode.addChild(dotnetNode);
    const dotnetCapNode = new teamsfx_api_1.QTreeNode(question_3.createCapabilityForDotNet());
    dotnetNode.addChild(dotnetCapNode);
    const triggerNodeRes = await question_2.getNotificationTriggerQuestionNode(inputs);
    if (triggerNodeRes.isErr())
        return teamsfx_api_1.err(triggerNodeRes.error);
    if (triggerNodeRes.value) {
        dotnetCapNode.addChild(triggerNodeRes.value);
    }
    const spfxNode = await spfx_1.getSPFxScaffoldQuestion(inputs.platform);
    if (spfxNode) {
        spfxNode.condition = { equals: constants_1.TabSPFxItem().id };
        dotnetCapNode.addChild(spfxNode);
    }
    dotnetCapNode.addChild(new teamsfx_api_1.QTreeNode(question_3.ProgrammingLanguageQuestionForDotNet));
    // only CLI need folder input
    if (teamsfx_api_1.CLIPlatforms.includes(inputs.platform)) {
        runtimeNode.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
    }
    runtimeNode.addChild(new teamsfx_api_1.QTreeNode(question_3.createAppNameQuestion()));
    return teamsfx_api_1.ok(runtimeNode.trim());
}
async function getQuestionsForCreateProjectInVSC(inputs) {
    var _a, _b, _c, _d, _e, _f;
    if (inputs[question_3.CoreQuestionNames.CreateFromScratch] === question_3.ScratchOptionNoVSC().id) {
        // Create from sample flow
        const sampleNode = new teamsfx_api_1.QTreeNode(question_3.SampleSelect());
        sampleNode.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
        return teamsfx_api_1.ok(sampleNode.trim());
    }
    // We will always create a project from scratch in VSC.
    inputs[question_3.CoreQuestionNames.CreateFromScratch] = question_3.ScratchOptionYesVSC().id;
    if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
        inputs[question_3.CoreQuestionNames.ProjectType] =
            (_a = inputs[question_3.CoreQuestionNames.ProjectType]) !== null && _a !== void 0 ? _a : (_b = developerPortalScaffoldUtils_1.getTemplateId(inputs.teamsAppFromTdp)) === null || _b === void 0 ? void 0 : _b.projectType;
        inputs[question_3.CoreQuestionNames.Capabilities] =
            (_c = inputs[question_3.CoreQuestionNames.Capabilities]) !== null && _c !== void 0 ? _c : (_d = developerPortalScaffoldUtils_1.getTemplateId(inputs.teamsAppFromTdp)) === null || _d === void 0 ? void 0 : _d.templateId;
    }
    // create new project root
    const root = new teamsfx_api_1.QTreeNode({ type: "group" });
    // project type
    const capQuestion = question_3.createNewProjectQuestionWith2Layers(inputs);
    const typeNode = new teamsfx_api_1.QTreeNode(capQuestion);
    root.addChild(typeNode);
    // bot type capabilities
    const botTypeNode = new teamsfx_api_1.QTreeNode(question_3.getBotProjectQuestionNode(inputs));
    botTypeNode.condition = {
        equals: constants_1.NewProjectTypeBotOptionItem().id,
    };
    typeNode.addChild(botTypeNode);
    const triggerNodeRes = await question_2.getNotificationTriggerQuestionNode(inputs);
    if (triggerNodeRes.isErr())
        return teamsfx_api_1.err(triggerNodeRes.error);
    if (triggerNodeRes.value) {
        botTypeNode.addChild(triggerNodeRes.value);
    }
    // tab type
    const tabTypeNode = new teamsfx_api_1.QTreeNode(question_3.getTabTypeProjectQuestionNode(inputs));
    tabTypeNode.condition = {
        equals: constants_1.NewProjectTypeTabOptionItem().id,
    };
    typeNode.addChild(tabTypeNode);
    const spfxNode = await spfx_1.getSPFxScaffoldQuestion(inputs.platform);
    if (spfxNode) {
        spfxNode.condition = { equals: constants_1.TabSPFxItem().id };
        tabTypeNode.addChild(spfxNode);
    }
    // message extension type
    const messageExtensionTypeNode = new teamsfx_api_1.QTreeNode(question_3.getMessageExtensionTypeProjectQuestionNode(inputs));
    messageExtensionTypeNode.condition = {
        equals: constants_1.NewProjectTypeMessageExtensionOptionItem().id,
    };
    typeNode.addChild(messageExtensionTypeNode);
    // Outlook addin type
    const outlookAddinTypeNode = new teamsfx_api_1.QTreeNode(question_3.getOutlookAddinTypeProjectQuestionNode(inputs));
    outlookAddinTypeNode.condition = {
        equals: constants_1.NewProjectTypeOutlookAddinOptionItem().id,
    };
    typeNode.addChild(outlookAddinTypeNode);
    outlookAddinTypeNode.addChild(question_1.getQuestionsForScaffolding());
    // Language
    const programmingLanguage = new teamsfx_api_1.QTreeNode(question_3.ProgrammingLanguageQuestion);
    programmingLanguage.condition = {
        notEquals: constants_1.NewProjectTypeOutlookAddinOptionItem().id,
    };
    typeNode.addChild(programmingLanguage);
    root.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
    const defaultName = !((_e = inputs.teamsAppFromTdp) === null || _e === void 0 ? void 0 : _e.appName)
        ? undefined
        : utils_1.convertToAlphanumericOnly((_f = inputs.teamsAppFromTdp) === null || _f === void 0 ? void 0 : _f.appName);
    root.addChild(new teamsfx_api_1.QTreeNode(question_3.createAppNameQuestion(defaultName)));
    if (developerPortalScaffoldUtils_1.isFromDevPortal(inputs)) {
        const updateTabUrls = await getQuestionsForUpdateStaticTabUrls(inputs.teamsAppFromTdp);
        if (updateTabUrls) {
            typeNode.addChild(updateTabUrls);
        }
        const updateBotIds = await getQuestionsForUpdateBotIds(inputs.teamsAppFromTdp);
        if (updateBotIds) {
            typeNode.addChild(updateBotIds);
        }
    }
    return teamsfx_api_1.ok(root.trim());
}
async function getQuestionsForUpdateStaticTabUrls(appDefinition) {
    if (!utils_2.isPersonalApp(appDefinition)) {
        return undefined;
    }
    const updateTabUrls = new teamsfx_api_1.QTreeNode({ type: "group" });
    const tabs = appDefinition.staticTabs;
    const tabsWithContentUrls = tabs.filter((o) => !!o.contentUrl);
    const tabsWithWebsiteUrls = tabs.filter((o) => !!o.websiteUrl);
    if (tabsWithWebsiteUrls.length > 0) {
        updateTabUrls.addChild(new teamsfx_api_1.QTreeNode(question_3.tabsWebsitetUrlQuestion(tabsWithWebsiteUrls)));
    }
    if (tabsWithContentUrls.length > 0) {
        updateTabUrls.addChild(new teamsfx_api_1.QTreeNode(question_3.tabsContentUrlQuestion(tabsWithContentUrls)));
    }
    return updateTabUrls;
}
async function getQuestionsForUpdateBotIds(appDefinition) {
    if (!utils_2.needBotCode(appDefinition)) {
        return undefined;
    }
    const bots = appDefinition.bots;
    const messageExtensions = appDefinition.messagingExtensions;
    // can add only one bot. If existing, the length is 1.
    const botId = !!bots && bots.length > 0 ? bots[0].botId : undefined;
    // can add only one message extension. If existing, the length is 1.
    const messageExtensionId = !!messageExtensions && messageExtensions.length > 0 ? messageExtensions[0].botId : undefined;
    return new teamsfx_api_1.QTreeNode(question_3.BotIdsQuestion(botId, messageExtensionId));
}
async function getQuestionsForCreateProjectV2(inputs) {
    if (featureFlags_1.isCLIDotNetEnabled() && teamsfx_api_1.CLIPlatforms.includes(inputs.platform)) {
        return getQuestionsForCreateProjectWithDotNet(inputs);
    }
    else if (inputs.platform === teamsfx_api_1.Platform.VSCode) {
        return getQuestionsForCreateProjectInVSC(inputs);
    }
    else {
        return getQuestionsForCreateProjectWithoutDotNet(inputs);
    }
}
exports.getQuestionsForCreateProjectV2 = getQuestionsForCreateProjectV2;
function addOfficeAddinQuestions(node) {
    const createNewAddin = new teamsfx_api_1.QTreeNode({ type: "group" });
    createNewAddin.condition = { equals: question_3.CreateNewOfficeAddinOption().id };
    node.addChild(createNewAddin);
    const capNode = new teamsfx_api_1.QTreeNode(question_3.createCapabilityForOfficeAddin());
    createNewAddin.addChild(capNode);
    capNode.addChild(question_1.getQuestionsForScaffolding());
    createNewAddin.addChild(new teamsfx_api_1.QTreeNode(question_3.QuestionRootFolder()));
    createNewAddin.addChild(new teamsfx_api_1.QTreeNode(question_3.createAppNameQuestion()));
}
exports.addOfficeAddinQuestions = addOfficeAddinQuestions;
