// @microsoft/teamsfx-core/build/component/resource/botService/appStudio/appStudioClient.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppStudioClient = exports.handleBotFrameworkError = void 0;
const tslib_1 = require("tslib");
/**
 * @author Ivan Jobs <ruhe@microsoft.com>
 */
const IBotRegistration_1 = require("./interfaces/IBotRegistration");
const axios_1 = tslib_1.__importDefault(require("axios"));
const errors_1 = require("../errors");
const strings_1 = require("../strings");
const retryHandler_1 = require("../retryHandler");
const messages_1 = require("../messages");
const constants_1 = require("../../appManifest/constants");
const error_1 = require("../../../error");
const result_1 = require("../result");
const appStudioClient_1 = require("../../appManifest/appStudioClient");
const common_1 = require("../common");
const commonConstant_1 = require("../../../constant/commonConstant");
const constants_2 = require("../constants");
function handleBotFrameworkError(e, apiName) {
    var _a, _b, _c, _d;
    if (((_a = e.response) === null || _a === void 0 ? void 0 : _a.status) === commonConstant_1.HttpStatusCode.NOTFOUND) {
        return undefined; // Stands for NotFound.
    }
    else if (((_b = e.response) === null || _b === void 0 ? void 0 : _b.status) === commonConstant_1.HttpStatusCode.UNAUTHORIZED) {
        throw new errors_1.BotFrameworkNotAllowedToAcquireTokenError();
    }
    else if (((_c = e.response) === null || _c === void 0 ? void 0 : _c.status) === commonConstant_1.HttpStatusCode.FORBIDDEN) {
        throw new errors_1.BotFrameworkForbiddenResultError();
    }
    else if (((_d = e.response) === null || _d === void 0 ? void 0 : _d.status) === commonConstant_1.HttpStatusCode.TOOMANYREQS) {
        throw new errors_1.BotFrameworkConflictResultError();
    }
    else {
        e.teamsfxUrlName = constants_2.TeamsFxUrlNames[apiName];
        throw appStudioClient_1.AppStudioClient.wrapException(e, apiName);
    }
}
exports.handleBotFrameworkError = handleBotFrameworkError;
class AppStudioClient {
    static newAxiosInstance(accessToken) {
        accessToken = error_1.CheckThrowSomethingMissing(result_1.FxBotPluginResultFactory.source, strings_1.ConfigNames.APPSTUDIO_TOKEN, accessToken);
        const instance = axios_1.default.create({
            headers: {
                post: {
                    Authorization: `Bearer ${accessToken}`,
                    "Client-Source": "teamstoolkit",
                },
                get: {
                    Authorization: `Bearer ${accessToken}`,
                    "Client-Source": "teamstoolkit",
                },
            },
        });
        instance.interceptors.request.use(function (config) {
            config.params = Object.assign({ teamstoolkit: true }, config.params);
            return config;
        });
        return instance;
    }
    /**
     * Set user region
     * @param _region e.g. https://dev.teams.microsoft.com/amer
     */
    static setRegion(region) {
        AppStudioClient.baseUrl = region;
    }
    static async getBotRegistration(token, botId) {
        appStudioClient_1.AppStudioClient.sendStartEvent(constants_1.APP_STUDIO_API_NAMES.GET_BOT);
        const axiosInstance = AppStudioClient.newAxiosInstance(token);
        try {
            const response = await retryHandler_1.RetryHandler.Retry(() => axiosInstance.get(`${AppStudioClient.baseUrl}/api/botframework/${botId}`));
            if (common_1.isHappyResponse(response)) {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                appStudioClient_1.AppStudioClient.sendSuccessEvent(constants_1.APP_STUDIO_API_NAMES.GET_BOT);
                return response.data; // response cannot be undefined as it's checked in isHappyResponse.
            }
            else {
                // Defensive code and it should never reach here.
                throw new Error("Failed to get data");
            }
        }
        catch (e) {
            handleBotFrameworkError(e, constants_1.APP_STUDIO_API_NAMES.GET_BOT);
        }
    }
    static async createBotRegistration(token, registration, checkExistence = true, context) {
        var _a;
        appStudioClient_1.AppStudioClient.sendStartEvent(constants_1.APP_STUDIO_API_NAMES.CREATE_BOT);
        const axiosInstance = AppStudioClient.newAxiosInstance(token);
        if (registration.botId && checkExistence) {
            const botReg = await AppStudioClient.getBotRegistration(token, registration.botId);
            if (botReg) {
                (_a = context === null || context === void 0 ? void 0 : context.logProvider) === null || _a === void 0 ? void 0 : _a.info(messages_1.Messages.BotResourceExist("Appstudio"));
                appStudioClient_1.AppStudioClient.sendSuccessEvent(constants_1.APP_STUDIO_API_NAMES.CREATE_BOT);
                return;
            }
        }
        try {
            const response = await retryHandler_1.RetryHandler.Retry(() => axiosInstance.post(`${AppStudioClient.baseUrl}/api/botframework`, registration));
            if (!common_1.isHappyResponse(response)) {
                throw new errors_1.ProvisionError(strings_1.CommonStrings.APP_STUDIO_BOT_REGISTRATION);
            }
            appStudioClient_1.AppStudioClient.sendSuccessEvent(constants_1.APP_STUDIO_API_NAMES.CREATE_BOT);
        }
        catch (e) {
            handleBotFrameworkError(e, constants_1.APP_STUDIO_API_NAMES.CREATE_BOT);
        }
        return;
    }
    static async updateMessageEndpoint(token, botId, endpoint) {
        const botReg = await AppStudioClient.getBotRegistration(token, botId);
        if (!botReg) {
            throw new errors_1.BotRegistrationNotFoundError(botId);
        }
        botReg.messagingEndpoint = endpoint;
        if (botReg.configuredChannels === undefined || botReg.configuredChannels.length === 0) {
            botReg.configuredChannels = [IBotRegistration_1.BotChannelType.MicrosoftTeams];
        }
        await AppStudioClient.updateBotRegistration(token, botReg);
        return;
    }
    static async updateBotRegistration(token, botReg) {
        appStudioClient_1.AppStudioClient.sendStartEvent(constants_1.APP_STUDIO_API_NAMES.UPDATE_BOT);
        const axiosInstance = AppStudioClient.newAxiosInstance(token);
        try {
            const response = await retryHandler_1.RetryHandler.Retry(() => axiosInstance.post(`${AppStudioClient.baseUrl}/api/botframework/${botReg.botId}`, botReg));
            if (!common_1.isHappyResponse(response)) {
                throw new errors_1.ConfigUpdatingError(strings_1.ConfigNames.MESSAGE_ENDPOINT);
            }
            appStudioClient_1.AppStudioClient.sendSuccessEvent(constants_1.APP_STUDIO_API_NAMES.UPDATE_BOT);
        }
        catch (e) {
            handleBotFrameworkError(e, constants_1.APP_STUDIO_API_NAMES.UPDATE_BOT);
        }
        return;
    }
}
exports.AppStudioClient = AppStudioClient;
AppStudioClient.baseUrl = constants_1.getAppStudioEndpoint();
