// @microsoft/teamsfx-core/build/component/resource/botService/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeamsFxUrlNames = exports.TelemetryKeys = exports.MaxLengths = exports.Alias = exports.Links = exports.ErrorNames = exports.Retry = exports.AADRegistrationConstants = exports.RegularExprs = void 0;
const constants_1 = require("../appManifest/constants");
/**
 * @author zhijie <zhihuan@microsoft.com>
 */
class RegularExprs {
}
exports.RegularExprs = RegularExprs;
RegularExprs.CHARS_TO_BE_SKIPPED = /[^a-zA-Z\d]/g;
class AADRegistrationConstants {
}
exports.AADRegistrationConstants = AADRegistrationConstants;
AADRegistrationConstants.GRAPH_REST_BASE_URL = "https://graph.microsoft.com/v1.0";
AADRegistrationConstants.AZURE_AD_MULTIPLE_ORGS = "AzureADMultipleOrgs";
class Retry {
}
exports.Retry = Retry;
Retry.RETRY_TIMES = 6;
Retry.BACKOFF_TIME_MS = 5000;
class ErrorNames {
}
exports.ErrorNames = ErrorNames;
// System Exceptions
ErrorNames.PRECONDITION_ERROR = "PreconditionError";
ErrorNames.PROVISION_ERROR = "ProvisionError";
ErrorNames.CONFIG_UPDATING_ERROR = "ConfigUpdatingError";
ErrorNames.CONFIG_VALIDATION_ERROR = "ConfigValidationError";
ErrorNames.BOT_REGISTRATION_NOTFOUND_ERROR = "BotRegistrationNotFoundError";
ErrorNames.MSG_ENDPOINT_UPDATING_ERROR = "MessageEndpointUpdatingError";
ErrorNames.COMMAND_EXECUTION_ERROR = "CommandExecutionError";
ErrorNames.CALL_APPSTUDIO_API_ERROR = "CallAppStudioAPIError";
ErrorNames.CREATE_BOT_REGISTRATION_API_ERROR = "CreateBotRegistrationApiError";
// User Exceptions
ErrorNames.PACK_DIR_EXISTENCE_ERROR = "PackDirectoryExistenceError";
ErrorNames.ACQUIRE_BOT_FRAMEWORK_TOKEN_ERROR = "AcquireBotFrameworkTokenError";
ErrorNames.FORBIDDEN_RESULT_BOT_FRAMEWORK_ERROR = "ForbiddenResultBotFrameworkError";
ErrorNames.CONFLICT_RESULT_BOT_FRAMEWORK_ERROR = "ConflictResultBotFrameworkError";
class Links {
}
exports.Links = Links;
Links.ISSUE_LINK = "https://github.com/OfficeDev/TeamsFx/issues/new";
Links.HELP_LINK = "https://aka.ms/teamsfx-bot-help";
class Alias {
}
exports.Alias = Alias;
Alias.TEAMS_BOT_PLUGIN = "BT";
Alias.BICEP_MODULE = "bot";
class MaxLengths {
}
exports.MaxLengths = MaxLengths;
MaxLengths.AAD_DISPLAY_NAME = 120;
class TelemetryKeys {
}
exports.TelemetryKeys = TelemetryKeys;
TelemetryKeys.Component = "component";
TelemetryKeys.Success = "success";
TelemetryKeys.ErrorType = "error-type";
TelemetryKeys.ErrorMessage = "error-message";
TelemetryKeys.ErrorCode = "error-code";
TelemetryKeys.AppId = "appid";
TelemetryKeys.HostType = "bot-host-type";
TelemetryKeys.BotCapabilities = "bot-capabilities";
TelemetryKeys.StatusCode = "status-code";
TelemetryKeys.Url = "url";
TelemetryKeys.Method = "method";
exports.TeamsFxUrlNames = {
    [constants_1.APP_STUDIO_API_NAMES.CREATE_BOT]: "<create-bot-registration>",
    [constants_1.APP_STUDIO_API_NAMES.GET_BOT]: "<get-bot-registration>",
    [constants_1.APP_STUDIO_API_NAMES.UPDATE_BOT]: "<update-message-endpoint>",
};
