// @microsoft/teamsfx-core/build/component/configManager/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validator = void 0;
const tslib_1 = require("tslib");
const ajv_1 = tslib_1.__importDefault(require("ajv"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const path_1 = tslib_1.__importDefault(require("path"));
const folder_1 = require("../../folder");
const diagnostic_1 = require("./diagnostic");
const supportedVersions = ["1.0.0", "1.1.0"];
class Validator {
    constructor() {
        this.impl = new Map();
        for (const version of supportedVersions) {
            this.initVersion(version);
        }
    }
    initVersion(version) {
        const ajv = new ajv_1.default();
        ajv.addKeyword("deprecationMessage");
        const schemaPath = path_1.default.join(folder_1.getResourceFolder(), "yaml-schema", version, "yaml.schema.json");
        const schema = fs_extra_1.default.readJSONSync(schemaPath);
        const schemaString = fs_extra_1.default.readFileSync(schemaPath, "utf8");
        this.impl.set(version, {
            validator: ajv.compile(schema),
            yamlDiagnostics: new diagnostic_1.YAMLDiagnostics(schemaPath, schemaString),
        });
    }
    isVersionSupported(version) {
        return this.supportedVersions().includes(version);
    }
    supportedVersions() {
        return supportedVersions;
    }
    latestSupportedVersion() {
        return supportedVersions[supportedVersions.length - 1];
    }
    async generateDiagnosticMessage(yamlPath, yamlString, version) {
        const impl = this.impl.get(version !== null && version !== void 0 ? version : this.latestSupportedVersion());
        if (!impl) {
            return undefined;
        }
        return impl.yamlDiagnostics.doValidation(yamlPath, yamlString);
    }
    validate(obj, version) {
        const impl = this.impl.get(version !== null && version !== void 0 ? version : this.latestSupportedVersion());
        if (!impl) {
            return false;
        }
        return !!impl.validator(obj);
    }
}
exports.Validator = Validator;
